dex/src/main/java/io/crate/data/CompositeBatchIterator.java

```diff
@@ -202,15 +202,25 @@ public boolean moveNext() {
                     CompletableFuture<CompletableFuture> future = supplyAsync(() -> {
                         ArrayList<CompletableFuture<?>> futures = new ArrayList<>(batchIterators.size());
                         for (BatchIterator<T> batchIterator: batchIterators) {
-                            futures.add(batchIterator.loadNextBatch().toCompletableFuture());
+                            try {
+                                futures.add(batchIterator.loadNextBatch().toCompletableFuture());
+                            } catch (Throwable t) {
+                                return CompletableFuture.failedFuture(t);
+                            }
                         }
                         return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
                     }, executor);
                     nestedFutures.add(future);
                 }
             } else {
                 for (BatchIterator<T> iterator: itToLoad) {
-                    nestedFutures.add(supplyAsync(() -> iterator.loadNextBatch().toCompletableFuture(), executor));
+                    nestedFutures.add(supplyAsync(() -> {
+                        try {
+                            return iterator.loadNextBatch().toCompletableFuture();
+                        } catch (Throwable t) {
+                            return CompletableFuture.failedFuture(t);
```

Comment:
I guess this case is one of the downsides of the checked exception. `supplyAsync` would already handle the exception correctly :(